otherlibs/stdune/src/compact_position.ml

```diff
@@ -0,0 +1,134 @@
+module Position = struct
+  (* We encode the position in three, 21 bit fields: [cnum][lnum][bol] *)
+  type t = int
+
+  let field_size = 21
+  let mask_bol = (1 lsl field_size) - 1
+  let mask_lnum = mask_bol lsl field_size
+  let mask_cnum = mask_lnum lsl field_size
+
+  let small_enough =
+    let max_size = 1 lsl field_size in
+    let test int = int <= max_size in
+    fun [@inline] { Lexing.pos_bol; pos_cnum; pos_lnum; pos_fname = _ } ->
+      test pos_bol && test pos_cnum && test pos_lnum
+  ;;
+
+  let[@inline] of_position { Lexing.pos_bol; pos_cnum; pos_lnum; pos_fname = _ } =
```

Comment:
I think this way of defining the bit-twiddling is unnecessarily confusing. How about something like this?:

```ocaml
  let field_size = 21
  let field_mask = (1 lsl field_size) - 1
  let shift_bol = 0
  let shift_lnum = field_size
  let shift_cnum = 2 * field_size

  let[@inline] of_position { Lexing.pos_bol; pos_cnum; pos_lnum; pos_fname = _ } =
    ((pos_bol land field_mask) lsl shift_bol)
    lor ((pos_lnum land field_mask) lsl shift_lnum)
    lor ((pos_cnum land field_mask) lsl shift_cnum)
  ;;

  let[@inline] bol t = (t lsr shift_bol) land field_mask
  let[@inline] lnum t = (t lsr shift_lnum) land field_mask
  let[@inline] cnum t = (t lsr shift_cnum) land field_mask
```